### ISSUE TITLE
Operating motor with UART & setCurrent

### ISSUE THREAD

**Issue by huba80** [Chr157i4n/TMC2209_Raspberry_Pi#8]
My setup is that I would like to drive my 42HS48 motor with a TMC2209 v3.1 by UART.
The supplier who sold me the motor gave me a datasheet. According to this datasheet my motor is a 2A motor however I found other sellers online who sell the same motor but with different (lower) 1.68A current rating. Chinese sellers often dont have a clue of what they sell they just simply use a different specs that they find online and the specs sheet might have nothing to do with the                                                                                         physical motor in my hand. Im being cautious with the current rating.

Ive read the manual on the front page:

```
**test_script_05_vactual.py**
VACTUAL allows moving the motor by UART control. It gives the motor velocity in +-(2^23)-1 [μsteps / t]
```
If Im not mistaken this is the test file I need to run to test my motor movement using only UART.

Where can I set the maximum current within this file?

I found this:
 
```
#-----------------------------------------------------------------------
# these functions change settings in the TMC register
#-----------------------------------------------------------------------
tmc.setDirection_reg(False)
tmc.setVSense(True)
tmc.setCurrent(300)
tmc.setIScaleAnalog(True)
tmc.setInterpolation(True)
tmc.setSpreadCycle(False)
tmc.setMicrosteppingResolution(2)
tmc.setInternalRSense(False)
```
Here are my questions:

How does tmc.setCurrent(300) refer to the output current I want to register for the motor drive? What is this 300?

Because my motor's current rating is unknown Originally I set the Vref manually to 2.46V. This as the maximum for TMC2209 v3.1 using this calculator on the page:
https://wiki.fysetc.com/Silent2209/

`The function setCurrent only works correctly if the Vref voltage ist 1.2V, otherwise change Vref to 1.2V or give the actual Vref as parameter to the setCurrent function.`

Do I have to manually adjust the Vref for each driver if I want to use setCurrent?
How do I test if this set current is working? I dont want to set Vref too low for setcurrent 1.2V and if I cannot change by UART setting then driver burns out due to low current setting. Is there a way for me to get a report back if setCurrent works and at what value it is at?

Also do I need to connect DIAG pin somewhere to send a Vref voltage feedback to the RPi?

I dont not want to use microstepping. Just full step. Do I nee to change anything in:

`tmc.setMicrosteppingResolution(2)`

Thanks again.

**Comment by Chr157i4n** [comment]
what do you mean by
>  driver burns out due to low current setting

The setCurrent function only sets a current scale factor and the actual current still depends on the Vref, so either set the Vref to 1.2V and then use the setCurrent function with only the first Paramter (300). That is the current in mA.
Or use the setCurrent function with the Vref Parameter
```
setCurrent(300, Vref = 2.46)
```
You need to use a heatsink for medium current and for high currents you should use an active cooling fan.
Also you need to check the rating for your motor. 1A is more than enough for a Nema17 stepper.
I would suggest using a low current first and increase the current if the motor is loosing steps. Start at 200mA or 300mA and increase it by 100mA each time.
You should use the motor slightly higher than needed and not at the maximum current the driver/motor can withstand
\
As far as i know, there is no possibility to get the actual current over UART. If you find a way to get the actual current, pls let me know. 

\
the DIAG pin is for StallGuard detection.

\
\
for native full stepping you need to set
```
tmc.setMicrosteppingResolution(1)
```
then you might want to disable the internal interpolation of the TMC. That interpolates the fullstepping from the Raspberry Pi to 256 microstepping
```
tmc.setInterpolation(False)
```
But then the motor might vibrate. You can try what fits the best for your application.

Im also interested in your application. Could you please tell me what you are using the motor for.

**Comment by huba80** [comment]
Thank you.

The motor drive is for a feed screw drive pet feeder. Motor will operate intermittently only for 20 sec. After break for a few hours so there wont be a heavy use of the driver.

If I dont want to adjust the Vref pot just use UART setCurrent then how to change the value?

If I eg. want to have 1.5A output then:
setCurrent(300, Vref = 2.1)?

Or i need to change the 300 to something?

**Comment by Chr157i4n** [comment]
No, you need to measure the Vref you have set on the hardware driver and then put that value into the setCurrent function, after that you can use the first value to set the current in mA.
The setCurrent function can only scale down your current, so you need to set the Vref to as high as your maximum wanted current.

The first parameter of the setCurrent function is your desired current in mA.
The second paramet is your actual Vref voltage.

The calculation of the output current of the driver is rather complicated.
For high currents you need to set 
```
tmc.setVSense(False)
```
for smaller more accurate currents
```
tmc.setVSense(True)
```
should be used.
I should maybe implement automatic switching between those two.


Example: you want 1A
The FYSETC calculator sais you need 1.41V on Vref for that. You set as an example the Vref to 1.5V.
```
setVSense(False)
tmc.setCurrent(1000,Vref=1.5)
```
You get 
```
TMC2209_0: Vsense: 0
TMC2209_0: CS_IRun: 29
TMC2209_0: CS_IHold: 15
TMC2209_0: Delay: 10
TMC2209_0: ihold_irun
TMC2209_0: 0b10100001110100001111
TMC2209_0: writing ihold_irun
```

**CS_IRun** is the scaling factor while the motor is moving. Its a value between 0 and 31. 29 means the current is scaled down a little bit, because Vref (1.5V) is a little bit higher than needed (1.41V).

**CS_IHold** is the scaling factor when the motor is not moving, but the power still enabled.
In your application, you should disable the motor output between the movement.

\
\
Afterwards you can use the setCurrent function
```
tmc.setCurrent(500,Vref=1.5)
```
to get only 500mA without changing Vref again (here the CS_IRun is 14, approx the half of the value above)


For your Nema17 Motor i would suggest a current of 500-1000mA

**Comment by Chr157i4n** [comment]
i will also implement that I_scale_analog is used properly

> I_scale_analog (Reset default=1)
> 0: Use internal reference derived from 5VOUT
> 1: Use voltage supplied to VREF as current reference

Then the Voltage from Vref is not needed anymore to set the current with UART

**Comment by Chr157i4n** [comment]
ok, i implemented both:
its currently in the dev branch
https://github.com/Chr157i4n/TMC2209_Raspberry_Pi/tree/dev

- I_scale_analog so Vref is not needed anymore
- automatic switching between VSense for more convenience and more accurate current

So with that version you only need to use
```
tmc.setCurrent(1000)
```
to get 1000mA RMS current

Because the current scaling factor CS is an integer value between 0 and 31, you will not get exactly the set current. The scripts calculates and prints the approximate current for your wanted current
Example:
```
tmc.setCurrent(1000)
```
```
TMC2209_0: activated 5V-out for current scale
TMC2209_0: CS in range; using VSense False
TMC2209_0: activated Low sensitivity, high sense resistor voltage
TMC2209_0: CS_IRun: 17
TMC2209_0: CS_IHold: 9
TMC2209_0: Delay: 10
TMC2209_0: actual current: 994 mA
```
But that is only the resolution. I am not sure how accurate that is.

**Comment by huba80** [comment]
Thanks Christian, I dont understand half of this but Ill try to give it a try later today. :)
Might be a silly question but u mention 5V.
Now my driver is connected to Rpi 3.3V. should I change it and connect it to 5V?

**Comment by Chr157i4n** [comment]
> 0: Use internal reference derived from 5VOUT

Do you mean that? that is only for your information and that an TMC internal thing.
You dont need to change your wiring for that.

You have connected Vio to 3.3V? Vio should be 3-5V, so 3.3V in general is fine.
But the TMC has an internal 5V regulator, so when you connect the driver to VM (12V/24V) , there is no need of connecting anything to Vio

**Comment by huba80** [comment]
Im trying to make a small shield and I test two drivers at the same time. This TMC2209 v3.1 and GC6609 v1.0. I noticed that this GC6609 only worked then Vio was connected too. I just want both driver to work on the shield. TMC2209 worked without it only with GND.

**Comment by huba80** [comment]
I set setCurrent to 1000ma. Then I tried to run:

```
pi@raspberrypi:~/Desktop/TMCDEV $ sudo python3 -m test.test_script_05_vactual
/usr/bin/python3: No module named test.test_script_05_vactual
```
Why this works:

`sudo python3 -m tests.test_script_01_uart_connection`

But this one doesnt?
`sudo python3 -m test.test_script_05_vactual`

**Comment by Chr157i4n** [comment]
that is easy 😎
the folder of the tests is named **tests**
`sudo python3 -m tests.test_script_05_vactual`

**Comment by huba80** [comment]
Ohhh... gosh. I ddint realize the typo.

Now I ran the script:

```
pi@raspberrypi:~/Desktop/TMCDEV $ sudo python3 -m tests.test_script_05_vactual
---
SCRIPT START
---
TMC2209_0: Init
TMC2209_0: GPIO Init finished
TMC2209_0: Init finished
TMC2209_0: write normal motor direction
TMC2209_0: activated 5V-out for current scale
TMC2209_0: CS in range; using VSense False
TMC2209_0: activated Low sensitivity, high sense resistor voltage
TMC2209_0: CS_IRun: 17
TMC2209_0: CS_IHold: 9
TMC2209_0: Delay: 10
TMC2209_0: actual current: 994 mA
TMC2209_0: ihold_irun
TMC2209_0: 0b10100001000100001001
TMC2209_0: writing ihold_irun
TMC2209_0: activated Vref for current scale
TMC2209_0: writing microstep interpolation setting: True
TMC2209_0: activated Stealthchop
TMC2209_0: writing 1 microstep setting
TMC2209_0: writing MStep Reg Select: True
TMC2209_0: activated operation with external sense resistors
---
---
TMC2209_0: ---
TMC2209_0: INPUTS
TMC2209_0: 0b100001000000000000001001000001
TMC2209_0: spread is low
TMC2209_0: dir is high
TMC2209_0: step is low
TMC2209_0: en is high
TMC2209_0: ---
TMC2209_0: ---
TMC2209_0: CHOPPER CONTROL
TMC2209_0: 0b11000000000010000000001010011
TMC2209_0: native 1 microstep setting
TMC2209_0: interpolation to 256 microsteps
TMC2209_0: 0: Low sensitivity, high sense resistor voltage
TMC2209_0: ---
TMC2209_0: ---
TMC2209_0: DRIVER STATUS:
TMC2209_0: -0b111111111011110000000000000000
TMC2209_0: Info: motor is standing still
TMC2209_0: Info: motor is running on StealthChop
TMC2209_0: CS actual: 17
---
TMC2209_0: ---
TMC2209_0: GENERAL CONFIG
TMC2209_0: 0b110000001
TMC2209_0: Driver is using voltage supplied to VREF as current reference
TMC2209_0: Operation with external sense resistors
TMC2209_0: StealthChop PWM mode enabled
TMC2209_0: normal motor direction
TMC2209_0: INDEX shows the first microstep position of sequencer
TMC2209_0: INDEX output as selected by index_otpw
TMC2209_0: Microstep resolution selected by MSTEP register
TMC2209_0: ---
---
---
TMC2209_0: Motor output active: True
TMC2209_0: vactual
TMC2209_0: 0b110010000
TMC2209_0: writing vactual
TMC2209_0: vactual
TMC2209_0: 0b0
TMC2209_0: writing vactual
TMC2209_0: vactual
TMC2209_0: -0b110010000
TMC2209_0: writing vactual
TMC2209_0: Motor output active: False
---
---
TMC2209_0: Deinit
TMC2209_0: Motor output active: False
TMC2209_0: GPIO cleanup
---
SCRIPT FINISHED
---
```
For me this all seems o be fine. Because I burned out several stepper driver last week due t o RPi config error an wrong Vref setup I have not connected y stepper driver yet. I wish to avoid another burned out driver so Im sorry if Im being overly cautious and I have many questions. I just want to make sure that this time it will work out without little smoke. :)

What should I expect from the motor if I connect it to the driver and run again vactual script? Is it going to start to turn back and forth for some time?

**Comment by huba80** [comment]
Never mind. I run the script with motor connected. It worked fine. 
Could you help me where can I set the speed?
Is is "setVActual"?

**Comment by huba80** [comment]
I tried to increase

`tmc.setMaxSpeed(500)`

but it did nothing.

**Comment by Chr157i4n** [comment]
in the vactual test script the motor should move forward for 1 seconds then stop for 1 seconds and then move backwards for one second.

i have also destroyed not just one driver... one motor too

\

the parameter of VActual is the Speed in +-(2^23)-1 [μsteps / t]

in that case 
```
tmc.setAcceleration(2000)
tmc.setMaxSpeed(500)
```
those nothing

\
i also forgot to delete this line in all of the test scripts
```
tmc.setIScaleAnalog(True)
```

So please download the new version from the dev branch

**Comment by huba80** [comment]
Thanks again. This vactual setup worked me:

```
from src.TMC_2209.TMC_2209_StepperDriver import *
import time


print("---")
print("SCRIPT START")
print("---")





#-----------------------------------------------------------------------
# initiate the TMC_2209 class
# use your pins for pin_step, pin_dir, pin_en here
#-----------------------------------------------------------------------
tmc = TMC_2209(16, 27, 21)





#-----------------------------------------------------------------------
# set the loglevel of the libary (currently only printed)
# set whether the movement should be relative or absolute
# both optional
#-----------------------------------------------------------------------
tmc.setLoglevel(Loglevel.debug)
tmc.setMovementAbsRel(MovementAbsRel.absolute)





#-----------------------------------------------------------------------
# these functions change settings in the TMC register
#-----------------------------------------------------------------------
tmc.setDirection_reg(False)
tmc.setCurrent(1000)
tmc.setInterpolation(True)
tmc.setSpreadCycle(False)
tmc.setMicrosteppingResolution(2)
tmc.setInternalRSense(False)


print("---\n---")





#-----------------------------------------------------------------------
# these functions read and print the current settings in the TMC register
#-----------------------------------------------------------------------
tmc.readIOIN()
tmc.readCHOPCONF()
tmc.readDRVSTATUS()
tmc.readGCONF()

print("---\n---")





#-----------------------------------------------------------------------
# activate the motor current output
#-----------------------------------------------------------------------
tmc.setMotorEnabled(True)





#-----------------------------------------------------------------------
# move the motor for 1 second forward, stop for 1 second
# and then move backwards for 1 second
#-----------------------------------------------------------------------
tmc.setVActual(2400)
time.sleep(5)
tmc.setVActual(0)
time.sleep(1)
tmc.setVActual(-2400)
time.sleep(5)





#-----------------------------------------------------------------------
# deactivate the motor current output
#-----------------------------------------------------------------------
tmc.setMotorEnabled(False)

print("---\n---")





#-----------------------------------------------------------------------
# deinitiate the TMC_2209 class
#-----------------------------------------------------------------------
del tmc

print("---")
print("SCRIPT FINISHED")
print("---")
```
I would increase speed over 2400 but the motor is already shaky/noisy.
I assume that with this tmc.setCurrent(1000) setting the stepper driver completely disregards the Vref pot setup and it sets the current to the requested level ~1000ma. I no longer have to bother with the Vref pot.
So far 1000ma seems sufficient but Ill need to make a few more test. Driver didnt even warm up however I just ran a 2x 40 sec test.

**Comment by huba80** [comment]
I also tested S6609 stepper driver with UART connection and vactual tests.

On paper the result seems fine:

```
pi@raspberrypi:~/Desktop/TMCDEV $ sudo python3 -m tests.test_script_01_uart_connection
---
SCRIPT START
---
TMC2209_0: Init
TMC2209_0: GPIO Init finished
TMC2209_0: Init finished
TMC2209_0: write normal motor direction
TMC2209_0: activated 5V-out for current scale
TMC2209_0: CS too low; switching to VSense True
TMC2209_0: activated High sensitivity, low sense resistor voltage
TMC2209_0: CS_IRun: 9
TMC2209_0: CS_IHold: 4
TMC2209_0: Delay: 10
TMC2209_0: actual current: 306 mA
TMC2209_0: ihold_irun
TMC2209_0: 0b10100000100100000100
TMC2209_0: writing ihold_irun
TMC2209_0: activated Vref for current scale
TMC2209_0: writing microstep interpolation setting: True
TMC2209_0: activated Stealthchop
TMC2209_0: writing 2 microstep setting
TMC2209_0: writing MStep Reg Select: True
TMC2209_0: activated operation with external sense resistors
---
---
TMC2209_0: ---
TMC2209_0: INPUTS
TMC2209_0: 0b100000000000000000001111000001
TMC2209_0: spread is high
TMC2209_0: dir is high
TMC2209_0: step is high
TMC2209_0: en is high
TMC2209_0: ---
TMC2209_0: ---
TMC2209_0: CHOPPER CONTROL
TMC2209_0: -0b101000111111001111111110101101
TMC2209_0: native 2 microstep setting
TMC2209_0: interpolation to 256 microsteps
TMC2209_0: 1: High sensitivity, low sense resistor voltage
TMC2209_0: ---
TMC2209_0: ---
TMC2209_0: DRIVER STATUS:
TMC2209_0: 0b1000000000010010000000000000000
TMC2209_0: Info: motor is running
TMC2209_0: Info: motor is running on StealthChop
TMC2209_0: CS actual: 9
---
TMC2209_0: ---
TMC2209_0: GENERAL CONFIG
TMC2209_0: 0b110000001
TMC2209_0: Driver is using voltage supplied to VREF as current reference
TMC2209_0: Operation with external sense resistors
TMC2209_0: StealthChop PWM mode enabled
TMC2209_0: normal motor direction
TMC2209_0: INDEX shows the first microstep position of sequencer
TMC2209_0: INDEX output as selected by index_otpw
TMC2209_0: Microstep resolution selected by MSTEP register
TMC2209_0: ---
---
---
TMC2209_0: Deinit
TMC2209_0: Motor output active: False
TMC2209_0: GPIO cleanup
---
SCRIPT FINISHED
---
```
```
pi@raspberrypi:~/Desktop/TMCDEV $ sudo python3 -m tests.test_script_05_vactual
---
SCRIPT START
---
TMC2209_0: Init
TMC2209_0: GPIO Init finished
TMC2209_0: Init finished
TMC2209_0: write normal motor direction
TMC2209_0: activated 5V-out for current scale
TMC2209_0: CS in range; using VSense False
TMC2209_0: activated Low sensitivity, high sense resistor voltage
TMC2209_0: CS_IRun: 17
TMC2209_0: CS_IHold: 9
TMC2209_0: Delay: 10
TMC2209_0: actual current: 994 mA
TMC2209_0: ihold_irun
TMC2209_0: 0b10100001000100001001
TMC2209_0: writing ihold_irun
TMC2209_0: writing microstep interpolation setting: True
TMC2209_0: activated Stealthchop
TMC2209_0: writing 2 microstep setting
TMC2209_0: writing MStep Reg Select: True
TMC2209_0: activated operation with external sense resistors
---
---
TMC2209_0: ---
TMC2209_0: INPUTS
TMC2209_0: 0b100000000000000000001111000001
TMC2209_0: spread is high
TMC2209_0: dir is high
TMC2209_0: step is high
TMC2209_0: en is high
TMC2209_0: ---
TMC2209_0: ---
TMC2209_0: CHOPPER CONTROL
TMC2209_0: -0b101000111111101111111110101101
TMC2209_0: native 2 microstep setting
TMC2209_0: interpolation to 256 microsteps
TMC2209_0: 0: Low sensitivity, high sense resistor voltage
TMC2209_0: ---
TMC2209_0: ---
TMC2209_0: DRIVER STATUS:
TMC2209_0: 0b1000000000100010000000000000000
TMC2209_0: Info: motor is running
TMC2209_0: Info: motor is running on StealthChop
TMC2209_0: CS actual: 17
---
TMC2209_0: ---
TMC2209_0: GENERAL CONFIG
TMC2209_0: 0b110000000
TMC2209_0: Driver is using internal reference derived from 5VOUT
TMC2209_0: Operation with external sense resistors
TMC2209_0: StealthChop PWM mode enabled
TMC2209_0: normal motor direction
TMC2209_0: INDEX shows the first microstep position of sequencer
TMC2209_0: INDEX output as selected by index_otpw
TMC2209_0: Microstep resolution selected by MSTEP register
TMC2209_0: ---
---
---
TMC2209_0: Motor output active: True
TMC2209_0: vactual
TMC2209_0: 0b100101100000
TMC2209_0: writing vactual
TMC2209_0: vactual
TMC2209_0: 0b0
TMC2209_0: writing vactual
TMC2209_0: vactual
TMC2209_0: -0b100101100000
TMC2209_0: writing vactual
TMC2209_0: Motor output active: False
---
---
TMC2209_0: Deinit
TMC2209_0: Motor output active: False
TMC2209_0: GPIO cleanup
---
SCRIPT FINISHED
---

```
However the motor didnt move nothing. :( 
On the hackaday blog they say S6609 is a TMC2209 clone.

**Comment by Chr157i4n** [comment]
As you said earlier the S6609 clone doesnt have an internal 5V regulator either.
A documentation of that chip would be helpful, but i guess that functionality is simply not implemented in that chip.

Btw.
Parameter / Symbol | Unit | calculation / description / comment
-- | -- | --
fCLK[Hz] | [Hz] | clock frequency of the TMC2209 in [Hz]
μstep velocity v[Hz] | μsteps / s | v[Hz] = VACTUAL[2209] * ( fCLK[Hz]/2 / 2^23 ) <br /> With internal oscillator: <br /> v[Hz] = VACTUAL[2209] * 0.715Hz

Well, that explains why it only made around 0.715 turn instead of a full revolution...

So the VActual value for 1 revolution per second in fullstepping is not 200, but 200/0.715 ≈ 280

**Comment by huba80** [comment]
Christian. Thanks again and Ive sent you an emal.

**Comment by huba80** [comment]
How do you know it has made 0.715 turn? 
You are referring to something that I need to change in my config?

**Comment by Chr157i4n** [comment]
i let it turn with a vactual of 200 in fullstepping for 1 seconds and i noticed that it made only around 2/3 revolution.
The value 0.715 is from the TMC2209 datasheet and that fits my observation.

**Comment by Chr157i4n** [comment]
i have added a function: `setVActual_rps(rps)`
https://github.com/Chr157i4n/TMC2209_Raspberry_Pi/blob/837b194bbce62d08423bfe8ed4c3f058f5a02bc8/src/TMC_2209/TMC_2209_StepperDriver.py#L741-L743
With that function you can set the speed in revolutions per second

**Comment by huba80** [comment]
Thanks. Ill give it a try later on today.
Where should I add the expected RPM value in these lines?
Can you give an example?
I also use a gear ratio reducer mounted my stepper motor but regardless the RPM setting would be an improvement for me.

**Comment by Chr157i4n** [comment]
```
tmc.setVActual_rps(1)
```
you can use that line to get the motor to 1 rps (Revolutions per Second)

**Comment by huba80** [comment]
Thanks. Ive been busy in the past few days but now I thought Ill give a try to stallguard function. Ive run this:

`sudo python3 -m tests.test_script_04_stallguard`

Ive got this result:

```
pi@raspberrypi:~/Desktop/TMCDEV $ sudo python3 -m tests.test_script_04_stallguard
---
SCRIPT START
---
TMC2209_0: Init
TMC2209_0: GPIO Init finished
TMC2209_0: Init finished
TMC2209_0: write normal motor direction
TMC2209_0: activated 5V-out for current scale
TMC2209_0: CS in range; using VSense False
TMC2209_0: activated Low sensitivity, high sense resistor voltage
TMC2209_0: CS_IRun: 17
TMC2209_0: CS_IHold: 9
TMC2209_0: Delay: 10
TMC2209_0: actual current: 994 mA
TMC2209_0: ihold_irun
TMC2209_0: 0b10100001000100001001
TMC2209_0: writing ihold_irun
TMC2209_0: activated Vref for current scale
TMC2209_0: writing microstep interpolation setting: True
TMC2209_0: activated Stealthchop
TMC2209_0: writing 2 microstep setting
TMC2209_0: writing MStep Reg Select: True
TMC2209_0: activated operation with external sense resistors
---
---
TMC2209_0: ---
TMC2209_0: INPUTS
TMC2209_0: 0b100001000000000000000001010001
TMC2209_0: spread is low
TMC2209_0: dir is low
TMC2209_0: step is low
TMC2209_0: en is high
TMC2209_0: ---
TMC2209_0: ---
TMC2209_0: CHOPPER CONTROL
TMC2209_0: 0b10111000000010000000001010011
TMC2209_0: native 2 microstep setting
TMC2209_0: interpolation to 256 microsteps
TMC2209_0: 0: Low sensitivity, high sense resistor voltage
TMC2209_0: ---
TMC2209_0: ---
TMC2209_0: DRIVER STATUS:
TMC2209_0: 0b1000000000100010000000000000000
TMC2209_0: Info: motor is running
TMC2209_0: Info: motor is running on StealthChop
TMC2209_0: CS actual: 17
---
TMC2209_0: ---
TMC2209_0: GENERAL CONFIG
TMC2209_0: 0b110000001
TMC2209_0: Driver is using voltage supplied to VREF as current reference
TMC2209_0: Operation with external sense resistors
TMC2209_0: StealthChop PWM mode enabled
TMC2209_0: normal motor direction
TMC2209_0: INDEX shows the first microstep position of sequencer
TMC2209_0: INDEX output as selected by index_otpw
TMC2209_0: Microstep resolution selected by MSTEP register
TMC2209_0: ---
---
---
TMC2209_0: Motor output active: True
TMC2209_0: sgthrs
TMC2209_0: 0b110010
TMC2209_0: writing sgthrs
TMC2209_0: tcoolthrs
TMC2209_0: 0b11111010000
TMC2209_0: writing tcoolthrs
TMC2209_0: setup stallguard callback
Movement finished successfully
TMC2209_0: Motor output active: False
---
---
TMC2209_0: Deinit
TMC2209_0: Motor output active: False
TMC2209_0: GPIO cleanup
---
SCRIPT FINISHED
---
```

Im not sure if this is the latest updated stallguard file but it didint seem to work.
The motor made some noise when script started but it didnt really move. I assume that this script wanted to use PIN and DIR pins instead of UART?
PIN and DIR pins are not connected for me. Also which pin should I use for DIAG pin? GPIO26?

After this Ill try the REV/sec setting.

**Comment by Chr157i4n** [comment]
Yes, the test script for StallGuard is using STEP/DIR instead of VActual for moving the motor, but it should be possible to use VActual with StallGuard

```
rmc.runToPositionSteps(4000, MovementAbsRel.relative)
```
The runToPositionSteps function is using STEP/DIR and you need to use something like this instead
```
tmc.setVActual_rps(1)
time.sleep(1)
tmc.setVActual(0)
```
I will implement a function that lets you move the motor with VActual for a certain time or a certain distance, but that will not be very precise.

The DIAG pin can in theory be connected to any GPIO of the Raspberry Pi. The TMC is sending a signal on that Pin when StallGuard detects a stall. The test script is currently using GPIO 26 for that:
```
tmc.setStallguard_Callback(26, 50, my_callback)

# 1. param: pin connected to the tmc DIAG output
# 2. param: is the threshold StallGuard
# 3. param: is the callback function (threaded)
```

depending on your setup and current, you might need to test the 2. param (Threshold for StallGuard detection)

**Comment by huba80** [comment]
In my case accurate move is not important. I just need a fixed speed setting and time where I can determine the duration of the motor turn.
Stall guard is also important as it would give an indication that my feed screw jammed at some point so I can take an appropriate action.
So my best wish is to run it by UART only with stall-guard.

**Comment by huba80** [comment]
I tried something but I think my "hack" is completely wrong.:)

```
tmc.setStallguard_Callback(17 50, my_callback) # after this function call, StallGuard is active

finishedsuccessfully = tmc.setVActual_rps(1) time.sleep(1) tmc.setVActual(0)    #move 4000 steps forward

if(finishedsuccessfully == True):
    print("Movement finished successfully")
else:
    print("Movement was not completed")
```

Result:

```
pi@raspberrypi:~/Desktop/TMC $ sudo python3 -m tests.test_script_04_stallguard
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 188, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/usr/lib/python3.9/runpy.py", line 158, in _get_module_details
    code = loader.get_code(mod_name)
  File "<frozen importlib._bootstrap_external>", line 923, in get_code
  File "<frozen importlib._bootstrap_external>", line 853, in source_to_code
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/pi/TMC/tests/test_script_04_stallguard.py", line 96
    tmc.setStallguard_Callback(17 50, my_callback) # after this function call, StallGuard is active
                                  ^
SyntaxError: invalid syntax
```

**Comment by Chr157i4n** [comment]
there needs to be a comma between the paramters of a function:
`tmc.setStallguard_Callback(17, 50, my_callback)`

that will not work, as every command needs to be in a new line.
`finishedsuccessfully = tmc.setVActual_rps(1) time.sleep(1) tmc.setVActual(0)    #move 4000 steps forward`

i have updated the library and added the functionality to use StallGuard with VActual.
Update your version and have a look at [tests/test_script_04_stallguard.py](https://github.com/Chr157i4n/TMC2209_Raspberry_Pi/blob/main/tests/test_script_04_stallguard.py):

```
def my_callback(channel):  
    print("StallGuard!")
    tmc.stop()

tmc.setStallguard_Callback(26, 50, my_callback) # after this function call, StallGuard is active



finishedsuccessfully = tmc.runToPositionSteps(4000, MovementAbsRel.relative)        #uses STEP/DIR to move the motor
# finishedsuccessfully = tmc.setVActual_rpm(30, revolutions=10)                     #uses VActual Register to  move the motor


if(finishedsuccessfully == True):
    print("Movement finished successfully")
else:
    print("Movement was not completed")
```

there you can use the setVActual_rpm line instead of the runToPositionSteps line to use VActual with StallGuard
```
finishedsuccessfully = tmc.runToPositionSteps(4000, MovementAbsRel.relative)        #uses STEP/DIR to move the motor
# finishedsuccessfully = tmc.setVActual_rpm(30, revolutions=10)                     #uses VActual Register to  move the motor
```

**Comment by Chr157i4n** [comment]
i have also added the functionality to drive the motor better with setVActual_rpm:

drive the motor with 30 rpm for 4 seconds forward:
```
tmc.setVActual_rpm(30, duration=4)
```

drive the motor with 30 rpm 2 revolutions backwards:
```
tmc.setVActual_rpm(-30, revolutions=2)
```

But as i said earlier, this is not very accurate movement

**Comment by huba80** [comment]
Something is not quite right here.
Before testing the new file I started RPI and I heard a very smooth motor buzzing noise. Like the motor was trying to turn but couldnt.
This was straight after boot before even starting the terminal. I put my finger on the driver and was heating up very fast.
I didint understand why. When I quickly disconnected EN pin then it this odd behaviour stopped.
After that the code no longer managed to run the motor. I had to change EN pin to a different one. Then I could test again with the new EN pin.

Now same thing happened. I started to run the new stall guard code... motor started to make the same struggling after hitting the ENTER (no turn at all but I heard some quiet motor buzz) and nothing happened during stall guard code.

Driver started to heat up again so I quickly unplugged the driver from PSU.

I exit the terminal and plugged the PSU back again. As soon as PSU was connected back I heard the same motor buzzing (still no turn). I restarted RPi. 
Straight after RPi reboot I plugged VM pin back to PSU and driver blown up immediately. I started no code, no terminal.

Why is this happening? Why the motor struggles makes buzzing noise along with no turn when terminal was closed or after RPI reboot? I just dont understand.

Now I have another blown driver. :)

**Comment by huba80** [comment]
I found this in a forum:

`GPIOs 0-8, 14, and 15 will appear high at boot. The other GPIOs will appear low. So replace GPIO 4 with a low GPIO.`

This appears to be the TX RX pin.
Is it possible to turn GPIO pins off unless something turns them on? I just dont understand what controls the motor after RPI boot or even after when terminal window is closed. Very frustrating.

**Comment by Chr157i4n** [comment]
The driver enables the Motor current, when the signal on the EN pin is low. So in order to disable the Motor current output the signal on EN needs to be HIGH. So when you disconnect the EN pins the Motor will not turn at all.

disconnecting the PSU while the driver is enabled and outputting current is not good and may destroy the driver. disconnecting the EN pin is always better.

I dont know why your Motor would not turn and the driver was heating up really fast. Check the current you have set for the driver in your script.

**Comment by huba80** [comment]
I did not adjust the Vref with that driver which has just blown. I used GPIO 6 for Enabled.

Before starting the new script RPi booted ok with driver and EN connected to it.

After I started the new stall guard with 1000ma and changing hashtag.
When I was testing 1000ma has been perfectly fine. The driver didnt heat up at all. Barely got warm. 

The I started the code and motor started to make noise. My finger was on the driver. In like 5 sec it got extremely hot.

Can i set a GPIO pin off for boot?

**Comment by Chr157i4n** [comment]
> Before testing the new file I started RPI and I heard a very smooth motor buzzing noise. Like the motor was trying to turn but couldnt.

So did the problem (noisy, not turning motor) appear first after using the new script or before? You said that the motor was trying to turn (so the EN pins was on LOW level) after booting and before testing the new files. When exactly did the problem start? And if it was before using the new scripts, what did you change?

> I did not adjust the Vref with that driver which has just blown.

So that driver was a different one and the Vref should still be at 1.2V, the factory default?


https://roboticsbackend.com/raspberry-pi-gpios-default-state/#Code_to_test_GPIO_default_states
Maybe this sections helps: **Override the default state with an external pull up/down resistor**
connect a 10kOhm resistor between the pin you are using for EN and 3,3V. Then test with a multimeter which state the pin has after boot.

**Comment by huba80** [comment]
Ive just sent you and email again. I recorded a video to show you that something is really not right here.

**Comment by huba80** [comment]
Thanks. Here are my findings.
I added a 10K resistor between EN and 3.3V. I restarted RPi. It seemed ok.

This code also seemed to work:

```
tmc.setMotorEnabled(True)
tmc.setVActual(2300)
time.sleep(10)
tmc.setVActual(0)
tmc.setMotorEnabled(False)

time.sleep(10)

tmc.setMotorEnabled(True)
tmc.setVActual(-2300)
time.sleep(10)
tmc.setVActual(0)
tmc.setMotorEnabled(False)
```

Although this code:

```
tmc.setMotorEnabled(True)
tmc.setVActual(2300)
time.sleep(10)
tmc.setVActual(0)

time.sleep(10)

tmc.setVActual(-2300)
time.sleep(10)
tmc.setVActual(0)
tmc.setMotorEnabled(False)
```
Still let the motor run for an unknown reason. I hoped that this 10K pull up resistor would fix this problem too but it has not.
I wonder what drives the motor in its break.

Other findings regarding stallguard.

This code worked but not with full success:

```
pi@raspberrypi:~/Desktop/TMC $ sudo python3 -m tests.test_script_04_stallguard
---
SCRIPT START
---
TMC2209_0: Init
TMC2209_0: GPIO Init finished
TMC2209_0: Init finished
TMC2209_0: write normal motor direction
TMC2209_0: activated 5V-out for current scale
TMC2209_0: CS in range; using VSense False
TMC2209_0: activated Low sensitivity, high sense resistor voltage
TMC2209_0: CS_IRun: 17
TMC2209_0: CS_IHold: 9
TMC2209_0: Delay: 10
TMC2209_0: actual current: 994 mA
TMC2209_0: ihold_irun
TMC2209_0: 0b10100001000100001001
TMC2209_0: writing ihold_irun
TMC2209_0: writing microstep interpolation setting: True
TMC2209_0: activated Stealthchop
TMC2209_0: writing 2 microstep setting
TMC2209_0: writing MStep Reg Select: True
TMC2209_0: activated operation with external sense resistors
---
---
TMC2209_0: ---
TMC2209_0: INPUTS
TMC2209_0: 0b100001000000000000000011000001
TMC2209_0: spread is low
TMC2209_0: dir is low
TMC2209_0: step is high
TMC2209_0: en is high
TMC2209_0: ---
TMC2209_0: ---
TMC2209_0: CHOPPER CONTROL
TMC2209_0: 0b10111000000010000000001010011
TMC2209_0: native 2 microstep setting
TMC2209_0: interpolation to 256 microsteps
TMC2209_0: 0: Low sensitivity, high sense resistor voltage
TMC2209_0: ---
TMC2209_0: ---
TMC2209_0: DRIVER STATUS:
TMC2209_0: 0b1000000000100010000000000000000
TMC2209_0: Info: motor is running
TMC2209_0: Info: motor is running on StealthChop
TMC2209_0: CS actual: 17
---
TMC2209_0: ---
TMC2209_0: GENERAL CONFIG
TMC2209_0: 0b110000000
TMC2209_0: Driver is using internal reference derived from 5VOUT
TMC2209_0: Operation with external sense resistors
TMC2209_0: StealthChop PWM mode enabled
TMC2209_0: normal motor direction
TMC2209_0: INDEX shows the first microstep position of sequencer
TMC2209_0: INDEX output as selected by index_otpw
TMC2209_0: Microstep resolution selected by MSTEP register
TMC2209_0: ---
---
---
TMC2209_0: Motor output active: True
TMC2209_0: sgthrs
TMC2209_0: 0b110010
TMC2209_0: writing sgthrs
TMC2209_0: tcoolthrs
TMC2209_0: 0b11111010000
TMC2209_0: writing tcoolthrs
TMC2209_0: setup stallguard callback
TMC2209_0: vactual: 280 for 20.0 sec
TMC2209_0: 0b100011000
TMC2209_0: writing vactual
Movement finished successfully
TMC2209_0: ---
TMC2209_0: homing
TMC2209_0: activated Stealthchop
StallGuard!
TMC2209_0: Stallguard threshold:100
TMC2209_0: sgthrs
TMC2209_0: 0b1100100
TMC2209_0: writing sgthrs
TMC2209: writing not successful!
TMC2209: ifcnt: 71 0
TMC2209_0: tcoolthrs
TMC2209_0: 0b11111010000
TMC2209_0: writing tcoolthrs
TMC2209_0: setup stallguard callback
TMC2209_0: vactual: 280 for 2.0 sec
TMC2209_0: 0b100011000
TMC2209_0: writing vactual
TMC2209_0: ---
TMC2209_0: Motor output active: False
---
---
---
SCRIPT FINISHED
---
Fatal Python error: _enter_buffered_busy: could not acquire lock for <_io.BufferedWriter name='<stdout>'> at interpreter shutdown, possibly due to daemon threads
Python runtime state: finalizing (tstate=0xf26888)

Current thread 0xb6f09980 (most recent call first):
<no Python frame>
Aborted
```
I ren the code after again and that completed without error.
I dont understand all settings in the stallguard config.

In my stall-guard config I have threshold set to 50 but when the code runs it shows 100. Also how this threshold works? What is the minimum and maximum value and lower threshold allows more or less torque on the motor?

I tried to change it to 120 but then I felt that it is easier to get a "not completed" stall-guard error. 

Also in the stall-guard settings vactual uses this:

`finishedsuccessfully = tmc.setVActual_rpm(30, revolutions=10)                     #uses VActual Register to  move the motor`

Instead can I use a:

```
tmc.setVActual(2300)
time.sleep(10)
tmc.setVActual(0)
```
This would work better for me instead of revolution. In my case accuracy is not a priority.

Thanks

**Comment by Chr157i4n** [comment]
1. When the RPI boots, the GPIO have default values from the System
2. Your 10K Ohm pullup resistor between 3,3V and EN is overriding the system defaults. because the resistor value is lower.
3. When the scripts runs, tmc.setMotorEnabled(True) pulls the EN pin to GND and that overwrites your 10k resistor pullup.

So that your motor is still moving slowly while the motor is enabled and VActual is 0, will not be fixed by the 10k pullup resistor. That can only override the system default values for that pin and can not override the script setting the pin states.

When disabling the Motor output between moves works for your then use that.
\
Regardings StallGuard:
i have uploaded a new version 0.1.7. In your version i forgot to comment out the `tmc.doHoming(True, 50)` at the bottom and that had the arguments in the wrong order. So pls update again.
```
tmc.setVActual(2300)
time.sleep(10)
tmc.setVActual(0)
```
When you use that for the motor movement, the driver can detect a Stall with Stallguard, but the motor movement will not continue for the 10 seconds.
```
tmc.setVActual(2300, duration=10)
```
This will basicly do the same as your three line above, but then the motor can be stopped when detecting a stall.

You still need to call these lines before your movement to setup StallGuard:
```
def my_callback(channel):  
    print("StallGuard!")
    tmc.stop()

tmc.setStallguard_Callback(26, 50, my_callback) # after this function call, StallGuard is active
```
the `tmc.stop()` in the StallGuard callback does only work when you are using `tmc.setVActual(2300, duration=10)`
\
Regarding the Threshold for StallGuard:

> A higher value signals a lower motor load and more torque headroom.

So the higher you set your threshold, the sooner StallGuard will kick in. It should be a value between 0 and 256, but as you already noticed 120 is too high. You need to test that value for your Setup
